File: Backtracking/DFS/47Permutatiton2.py
# Same to 46 Permutation, but with duplicates in the list 所以需要检查有没有重复
from typing import List


class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 0: return [[]]
        res = []
        for _ in range(len(nums)):
            cur = nums.pop(0)
            perms = self.permuteUnique(nums)

            for i in perms:
                i.append(cur)
                if i not in res:
                    res.append(i)
            nums.append(cur)
        return res

File: Backtracking/DFS/47Permutatiton2.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        HashMap<Integer, Boolean> visited = new HashMap<>();

        for (int i = 0; i < nums.length; i++) {
            visited.put(i, false); // INDEX
        }

        dfs(nums, res, new ArrayList<>(), visited);
        return res;
    }

    public void dfs(int[] nums, List<List<Integer>> res, List<Integer> cur, HashMap<Integer, Boolean> visited) {
        if (nums.length == cur.size()) {
            res.add(new ArrayList<>(cur));
            return;
        }

        HashSet<Integer> set = new HashSet<>();
        for (int i = 0; i < nums.length; i++) {
            if (visited.get(i) == false) {
                if (set.contains(nums[i]))
                    continue;
                cur.add(nums[i]);
                visited.put(i, true);
                set.add(nums[i]);
                dfs(nums, res, cur, visited);
                cur.remove(cur.size() - 1);
                visited.put(i, false);
            }

        }
    }
}